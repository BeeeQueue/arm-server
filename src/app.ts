import process from "node:process"

import Fastify from "fastify"
import { customAlphabet, urlAlphabet } from "nanoid"

import Cors from "@fastify/cors"
import Helmet from "@fastify/helmet"

import pkgJson from "../package.json"
import { config } from "@/config"
import { docsPlugin } from "@/docs"
import { logger } from "@/lib/logger"
import { sendErrorToSentry } from "@/lib/sentry"
import { apiPlugin } from "@/routes/v1/ids/handler"
import { v2Plugin } from "@/routes/v2/ids/handler"
import { thetvdbPlugin } from "@/routes/v2/thetvdb/handler"
import { CacheTimes, cacheReply } from "@/utils"

const PROD = config.NODE_ENV === "production"

const nanoid = customAlphabet(urlAlphabet, 16)

export const buildApp = async () => {
  const App = Fastify({
    ignoreTrailingSlash: true,
    onProtoPoisoning: "remove",
    onConstructorPoisoning: "remove",
    trustProxy: PROD,
    genReqId: () => nanoid(),
    disableRequestLogging: process.env.NODE_ENV === "test",
    logger,
  })

  await App.register(Cors, {
    origin: true,
  })

  await App.register(Helmet, {
    hsts: false,
    contentSecurityPolicy: false,
  })

  App.addHook("onError", (request, reply, error, next) => {
    /* c8 ignore next 4 */
    if (error.validation != null) {
      if (request.method === "GET") {
        cacheReply(reply, CacheTimes.WEEK)
      }
    } else {
      sendErrorToSentry(error, request)
    }

    next()
  })

  await App.register(apiPlugin, { prefix: "/api" })
  await App.register(v2Plugin, { prefix: "/api/v2" })
  await App.register(thetvdbPlugin, { prefix: "/api/v2" })
  await App.register(docsPlugin)

  App.get("/", async (_, reply) => {
    cacheReply(reply, CacheTimes.WEEK * 2)
    void reply.redirect(301, pkgJson.homepage)
  })

  return App
}
