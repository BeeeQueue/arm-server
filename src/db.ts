import Knex from 'knex'
import { config } from '../knexfile'

const { NODE_ENV } = process.env
export const knex = Knex(
  config[NODE_ENV as 'development' | 'production' | 'test']
)

export interface Relation {
  anilist?: number
  anidb?: number
  myanimelist?: number
  kitsu?: number
}
