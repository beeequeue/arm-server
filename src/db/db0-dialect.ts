/* eslint-disable ts/consistent-type-imports */

import type { Connector, Database, Primitive } from "db0"
import {
	type CompiledQuery,
	type DatabaseConnection,
	type DatabaseIntrospector,
	type Dialect,
	type Driver,
	type Kysely,
	type QueryCompiler,
	type QueryResult,
	SqliteAdapter,
	SqliteIntrospector,
	SqliteQueryCompiler,
} from "kysely"

type SqliteDatabase = import("node:sqlite").DatabaseSync
// | import("better-sqlite3").DatabaseSync
// | import("sqlite3").DatabaseSync
// | import("@cloudflare/workers-types/latest").D1Database

/**
 * todo docs
 */
export class Db0SqliteDialect implements Dialect {
	#db: Database<Connector<SqliteDatabase>>

	constructor(db: Database<Connector<SqliteDatabase>>) {
		this.#db = db
	}

	createAdapter() {
		return new SqliteAdapter()
	}

	createDriver(): Driver {
		return new Db0SqliteDriver(this.#db)
	}

	createQueryCompiler(): QueryCompiler {
		return new SqliteQueryCompiler()
	}

	createIntrospector(db: Kysely<any>): DatabaseIntrospector {
		return new SqliteIntrospector(db)
	}
}

class Db0SqliteDriver implements Driver {
	#db: Database<Connector<SqliteDatabase>>

	constructor(db: Database<Connector<SqliteDatabase>>) {
		this.#db = db
	}

	async init(): Promise<void> {}

	async acquireConnection(): Promise<DatabaseConnection> {
		return new Db0SqliteConnection(this.#db)
	}

	async beginTransaction(conn: Db0SqliteConnection): Promise<void> {
		return conn.beginTransaction()
	}

	async commitTransaction(conn: Db0SqliteConnection): Promise<void> {
		return conn.commitTransaction()
	}

	async rollbackTransaction(conn: Db0SqliteConnection): Promise<void> {
		return conn.rollbackTransaction()
	}

	async releaseConnection(_conn: Db0SqliteConnection): Promise<void> {}

	async destroy(): Promise<void> {
		const instance = await this.#db.getInstance()

		// eslint-disable-next-line ts/no-unsafe-member-access
		if (typeof (instance as any).close === "function") {
			return (instance as never as { close: () => Promise<void> }).close()
		}
		// eslint-disable-next-line ts/no-unsafe-member-access
		if (typeof (instance as any).destroy === "function") {
			return (instance as never as { destroy: () => Promise<void> }).destroy()
		}
	}
}

class Db0SqliteConnection implements DatabaseConnection {
	#db: Database<Connector<SqliteDatabase>>

	// TODO: transactions?

	constructor(db: Database<Connector<SqliteDatabase>>) {
		if (db.dialect !== "sqlite") {
			throw new Error("db0 driver only supports sqlite dialects for now.")
		}

		this.#db = db
	}

	async executeQuery<O>(compiledQuery: CompiledQuery): Promise<QueryResult<O>> {
		// Transactions are not supported yet.
		// if (this.#transactionClient) return this.#transactionClient.executeQuery(compiledQuery)

		const results = await this.#db
			.prepare(compiledQuery.sql)
			.bind(...(compiledQuery.parameters as Primitive[]))
			.all()
			.then((data) => ({ data, error: null }))
			.catch((error: Error) => ({ data: null, error }))
		if (results.error) {
			console.dir(results.error)
			throw new Error("Failed to execute query", { cause: results.error })
		}

		const numAffectedRows =
			results.data.length > 0 ? BigInt(results.data.length) : undefined

		return {
			// insertId:
			//   results.meta.last_row_id === undefined || results.meta.last_row_id === null
			//     ? undefined
			//     : BigInt(results.meta.last_row_id),
			rows: (results.data as O[]) ?? [],
			numAffectedRows,
		}
	}

	async beginTransaction() {
		// this.#transactionClient = this.#transactionClient ?? new PlanetScaleConnection(this.#config)
		// this.#transactionClient.#conn.execute('BEGIN')
		throw new Error("db0 driver does not support transactions yet.")
	}

	async commitTransaction() {
		// if (!this.#transactionClient) throw new Error('No transaction to commit')
		// this.#transactionClient.#conn.execute('COMMIT')
		// this.#transactionClient = undefined
		throw new Error("db0 driver does not support transactions yet.")
	}

	async rollbackTransaction() {
		// if (!this.#transactionClient) throw new Error('No transaction to rollback')
		// this.#transactionClient.#conn.execute('ROLLBACK')
		// this.#transactionClient = undefined
		throw new Error("db0 driver does not support transactions yet.")
	}

	async *streamQuery<O>(
		_compiledQuery: CompiledQuery,
		_chunkSize: number,
	): AsyncIterableIterator<QueryResult<O>> {
		throw new Error("db0 driver does not support streaming (yet?)")
	}
}
