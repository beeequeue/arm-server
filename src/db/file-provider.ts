import { readdirSync } from "node:fs"
import path from "node:path"
import { pathToFileURL } from "node:url"

import type { Migration, MigrationProvider } from "kysely"

const isMigration = (obj: unknown): obj is Migration =>
	typeof (obj as { up?: unknown })?.up === "function"

/**
 * Reads all migrations from a folder in node.js.
 *
 * ```ts
 * new FileMigrationProvider("path/to/migrations/folder")
 * ```
 */
export class ActuallyWorkingMigrationProvider implements MigrationProvider {
	readonly #migrationDirPath: string

	constructor(migrationDirPath: string) {
		this.#migrationDirPath = migrationDirPath
	}

	async getMigrations(): Promise<Record<string, Migration>> {
		const migrations: Record<string, Migration> = {}
		const files = readdirSync(this.#migrationDirPath)

		for (const fileName of files) {
			if (
				fileName.endsWith(".js") ||
				(fileName.endsWith(".ts") && !fileName.endsWith(".d.ts")) ||
				fileName.endsWith(".mjs") ||
				(fileName.endsWith(".mts") && !fileName.endsWith(".d.mts"))
			) {
				const filePath = pathToFileURL(
					path.join(this.#migrationDirPath, fileName),
				).toString()
				const migration = (await import(filePath)) as Migration | { default?: Migration }
				const migrationKey = fileName.substring(0, fileName.lastIndexOf("."))

				// Handle esModuleInterop export's `default` prop...
				if (isMigration((migration as { default?: Migration })?.default)) {
					migrations[migrationKey] = (migration as { default?: Migration }).default!
				} else if (isMigration(migration)) {
					migrations[migrationKey] = migration
				}
			}
		}

		return migrations
	}
}
