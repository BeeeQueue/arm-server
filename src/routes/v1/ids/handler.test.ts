import type { IncomingMessage, ServerResponse } from "node:http"

import type { FastifyInstance } from "fastify"
import type { RawServerDefault } from "fastify/types/utils"
import type { Logger } from "pino"
import { afterAll, afterEach, beforeAll, describe, expect, it } from "vitest"

import { buildApp } from "@/app"
import type { Relation } from "@/db"
import { knex, Source } from "@/db"

let id = 0
const createRelations = async <N extends number>(
  amount: N,
): Promise<N extends 1 ? Relation : Relation[]> => {
  const relations = Array.from({ length: amount }).map(() => ({
    anilist: id++,
    anidb: id++,
    kitsu: id++,
    myanimelist: id++,
  }))

  await knex.insert(relations).into("relations")

  if (amount === 1) {
    return relations[0] as never
  }

  return relations as never
}

let app: FastifyInstance<
  RawServerDefault,
  IncomingMessage,
  ServerResponse<IncomingMessage>,
  Logger
>
beforeAll(async () => {
  app = await buildApp()
})

afterEach(() => knex.delete().from("relations"))

afterAll(async () => {
  await Promise.all([app.close(), knex.destroy()])
})

describe("query params", () => {
  it("fetches relation correctly", async () => {
    const relation = await createRelations(1)

    const response = await app.inject().get("/api/ids").query({
      source: Source.AniList,
      id: relation.anilist!.toString(),
    })

    expect(response.json()).toStrictEqual(relation)
    expect(response.statusCode).toBe(200)
    expect(response.headers["content-type"]).toContain("application/json")
  })

  it("returns null when id doesn't exist", async () => {
    const response = await app.inject().get("/api/ids").query({
      source: Source.Kitsu,
      id: "404" as never,
    })

    expect(response.json()).toBe(null)
    expect(response.statusCode).toBe(200)
    expect(response.headers["content-type"]).toContain("application/json")
  })

  it("can return a partial response", async () => {
    const relation: Relation = {
      anidb: 1337,
      anilist: 1337,
      // TODO
      myanimelist: null!,
      kitsu: null!,
    }
    await knex.insert(relation).into("relations")

    const response = await app.inject().get("/api/ids").query({
      source: Source.AniList,
      id: relation.anilist!.toString(),
    })

    expect(response.json()).toStrictEqual(relation)
    expect(response.statusCode).toBe(200)
    expect(response.headers["content-type"]).toContain("application/json")
  })
})

describe("json body", () => {
  describe("object input", () => {
    it("gET fails with json body", async () => {
      const relations = await createRelations(4)

      const response = await app
        .inject()
        .get("/api/ids")
        .body({
          [Source.AniDB]: relations[0].anidb,
        })

      expect(response.json()).toMatchSnapshot()
      expect(response.statusCode).toBe(400)
      expect(response.headers["content-type"]).toContain("application/json")
    })

    it("fetches a single relation", async () => {
      const relations = await createRelations(4)

      const response = await app
        .inject()
        .post("/api/ids")
        .body({
          [Source.AniDB]: relations[0].anidb,
        })

      expect(response.json()).toStrictEqual(relations[0])
      expect(response.statusCode).toBe(200)
      expect(response.headers["content-type"]).toContain("application/json")
    })

    it("errors correctly on an empty object", async () => {
      await createRelations(4)

      const response = await app.inject().post("/api/ids").body({})

      expect(response.json()).toMatchSnapshot()
      expect(response.statusCode).toBe(400)
      expect(response.headers["content-type"]).toContain("application/json")
    })

    it("returns null if not found", async () => {
      await createRelations(4)

      const response = await app.inject().post("/api/ids").body({ anidb: 100_000 })

      expect(response.json()).toBe(null)
      expect(response.statusCode).toBe(200)
      expect(response.headers["content-type"]).toContain("application/json")
    })

    it("can return a partial response", async () => {
      const relation: Relation = {
        anidb: 1337,
        anilist: 1337,
        myanimelist: null as never,
        kitsu: null as never,
      }
      await knex.insert(relation).into("relations")

      const response = await app.inject().post("/api/ids").body({
        anilist: 1337,
      })

      expect(response.json()).toStrictEqual(relation)
      expect(response.statusCode).toBe(200)
      expect(response.headers["content-type"]).toContain("application/json")
    })
  })

  describe("array input", () => {
    it("fetches relations correctly", async () => {
      const relations = await createRelations(4)

      const body = [
        { [Source.AniDB]: relations[0].anidb },
        { [Source.AniList]: 1000 },
        { [Source.Kitsu]: relations[2].kitsu },
      ]

      const result = [relations[0], null, relations[2]]

      const response = await app.inject().post("/api/ids").body(body)

      expect(response.json()).toStrictEqual(result)
      expect(response.statusCode).toBe(200)
      expect(response.headers["content-type"]).toContain("application/json")
    })

    it("responds correctly on no finds", async () => {
      const body = [{ [Source.AniList]: 1000 }, { [Source.Kitsu]: 1000 }]

      const result = [null, null]

      const response = await app.inject().post("/api/ids").body(body)

      expect(response.json()).toStrictEqual(result)
      expect(response.statusCode).toBe(200)
      expect(response.headers["content-type"]).toContain("application/json")
    })

    it("requires at least one source", async () => {
      const body = [{}]

      const response = await app.inject().post("/api/ids").body(body)

      expect(response.json()).toMatchSnapshot()
      expect(response.statusCode).toBe(400)
      expect(response.headers["content-type"]).toContain("application/json")
    })
  })
})
