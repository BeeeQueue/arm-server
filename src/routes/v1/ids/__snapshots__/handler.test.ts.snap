// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`json body > array input > requires at least one source 1`] = `
{
  "error": {
    "issues": [
      {
        "code": "custom",
        "message": "At least one source is required.",
        "path": [
          0,
        ],
      },
    ],
    "name": "ZodError",
  },
  "success": false,
}
`;

exports[`json body > object input > errors correctly on an empty object 1`] = `
{
  "error": {
    "issues": [
      {
        "code": "custom",
        "message": "At least one source is required.",
        "path": [],
      },
    ],
    "name": "ZodError",
  },
  "success": false,
}
`;

exports[`json body > object input > gET fails with json body 1`] = `
{
  "error": {
    "issues": [
      {
        "code": "invalid_type",
        "expected": "'anilist' | 'anidb' | 'myanimelist' | 'kitsu'",
        "message": "Invalid source",
        "path": [
          "source",
        ],
        "received": "undefined",
      },
      {
        "code": "invalid_type",
        "expected": "number",
        "message": "Invalid ID",
        "path": [
          "id",
        ],
        "received": "nan",
      },
    ],
    "name": "ZodError",
  },
  "success": false,
}
`;
